walk.js version 1.25

DaveDubUK opened the pull request:
walk.js version 1.25
-------------------------

Release notes, June 2015

This version of the script has been extensively refactored; simplified, performance optimised and made far easier to read / understand. All known problems due to Interface bugs have been isolated and bug reports submitted.
I am not looking to get paid for this, I just want get the project synced and to get my work 'out there'. 


**Updated Transitions**
Compared to the current walk script (version 1.1, as found in hifi/examples), the only significant added functionality is the transitions, as already paid for by High Fidelity (worklist https://worklist.net/20186).


**Coding Style**
All points raised by Philip in the code review for the transitions worklist item have been addressed. Further changes in the spirit of Philip's comments and HiFi's coding standards have also been made. Philip's original comments are here: https://github.com/highfidelity/hifi/pull/3987/files#r22338645
    

**That horrible UI**
The old UI has been completely removed and replaced with a simple webWindow settings panel.


**Footstep sounds**
~~Playing footstep sounds (i.e. calling Audio.playSound repeatedly) continues to disrupt the script's smooth execution. This behaviour seems to be due to the varying execution time of Audio.playSound. A bug report, complete with short test script has been submitted here: https://worklist.net/20579~~ 
Fixed by re-using audio injectors and Howard's 'Set injector options from script' fix (https://github.com/highfidelity/hifi/pull/5203)


**Avatar rotation axis**
Rotating an avatar's joints causes the avatar to be displaced from his origin / correct rotation axis. This affects both the current and beta versions of the script (walk.js 1.1 and 1.25) and seems to be an Interface bug. A bug report, complete with test script and movie, has been submitted here: https://worklist.net/20575
  

**Major optimisations**
The following functionality in particular has been overhauled (optimised and drastically simplified): 
    - existing animation replay code 
    - existing animation storage code
    - existing locomotion state detection code (to test transitions)


**Avatar compatibility**
The animations are compatible with Ozan's reference avatar and work well with the Mixamo avatars. There has been a lot of progress in the alpha forums recently towards establishing Avatar creation workflows using open source editors. As a result, there is a (temporary) option in the settings that applies some simple pre-rotations to circumnavigate the current Blender joint orientation issues. However, as can be seen under discussion here: https://alphas.highfidelity.io/t/walk-js-new-avis-physics-and-turn-detection/5460/39 , problems with the standardisation of avatar's shoulder joints are not yet totally resolved.


I hope this brings the script up to HiFi standards - if not, please let me know what needs to be done. 

 - davedub

howard-stearns commented:
Hi Dave,
This looks terrific! Lots of great work there, and a lot to absorb. I'm pretty new here, and not as fast as some of the guys in doing PR reviews.  If it's alright with you, I'll post questions here as they come up, even if there's a good chance I might answer my own question later. I'm thinking that way there's less time between feedback items, and maybe less overall calendar time.

Looking great.
cheers,
-H

howard-stearns commented on examples/html/walkSettings.html:
```diff
@@ -0,0 +1,84 @@
+                        
+                        if (data.type == "update") {
+                            if (data.armsFree !== undefined) {
+                                elArmsFree.checked = data.armsFree == true;
```
Any reason not to just say elArmsFree.checked = data.armsFree?

DaveDubUK replied:
Hi Howard, and thanks for having a look at my script.

After checking if data.armsFree is defined, I used the data.armsFree == true expression to ensure that no matter what type or value data.armsFree was, the checkbox would always get a boolean setting.

It may be a little overcautious, but I did copy the idea over from the existing gridControls.html example included with edit.js.

Will await your advice on whether or not to remove.

howard-stearns replied:
Either way. My own preference would be simpler code, but that's just my taste. Your call.

howard-stearns commented on examples/html/walkSettings.html:
```diff
@@ -0,0 +1,84 @@
+
+        </div>
+    </body>
+</html>
```
This looks gorgeous in your branch, but resorts to being merely "functional" when I merge with current upstream master.  You might want to check with recent versions of, e.g., style.css and see if you're compatible.

DaveDubUK replied:
I use the edit.js css file, as I felt the settings dialog didn't really warrant it's own. I've just merged with upstream master locally and tested, image attached. (I also manually checked to ensure I'm using the current css file.)

Would you be able to give me some steps to reproduce?

![walk-script-settings-dialog](https://cloud.githubusercontent.com/assets/5994985/8321501/e23bb872-1a51-11e5-950f-5ca6460615af.png)

DaveDubUK replied:
It got me again! Interface seems to cache any css at startup, even though I supplied the newly updated css, it kept using the old one. 

Creating walkStyle.css now...

howard-stearns replied:
That picture is what I get as well, which is fine. I just didn't know if that was intentional relative to the pretty/fancy one that I saw in one version.

howard-stearns commented on examples/html/walkSettings.html:
```diff
@@ -0,0 +1,84 @@
+        <div class="grid-section">
+
+            <div id="entity-list-header">
+                <input type="button" id="power" value="Power" style="margin-left:68px; margin-top:10px"></button>
```
When merged (see comment for line 84), I can't tell when "power" is on or off. There's no visual distinction.

DaveDubUK replied:
I used the edit.js webWindow code as an example for the button. The buttons do not respond in edit.js, so I assumed this was ok.

I am happy to put in some user feedback / different button states. If I do so, I'll create a separate css file for the settings dialog at the same time.

howard-stearns replied:
Yeah, if you can, I think it would be good to have the feedback.

howard-stearns commented on examples/libraries/walkApi.js:
```diff
@@ -241,368 +503,417 @@ animation = (function() {
-                case this.SIDE_STEP:
+            // harmonics
+            targetAnimation.harmonics = {};
+            if (isDefined(sourceAnimation.harmonics)) {
```
I like the modularity of having library/walkApi.js separated from walk.js. But if walkAPI is truly meant to be a separately loadable library module, I think it should be loadable on its own. I.e., it should define everything it needs or include something that does.  So isDefined() would need to be defined here.

However, I'm not following why we have such a function at all. Why not just say:
   if (sourceAnimation.harmonics) {
or
  if (sourceAnimation && sourceAnimation.harmonics) {
or, if really necessary, and I don't think it is necessary in this case
  if (sourceAnimation && sourceAnimation.harmonics !== undefined) {

DaveDubUK replied:
I agree - this is not necessary and should be removed. This was part of some early code that I missed during the big refactor.

howard-stearns commented on examples/libraries/walkApi.js:
```diff
@@ -241,368 +503,417 @@ animation = (function() {
+            targetAnimation.filters = {};
+            for (i in sourceAnimation.filters) {
+                // are any filters specified for this joint?
+                if (isDefined(sourceAnimation.filters[i])) {
```
See isDefined comment above.

howard-stearns commented on examples/libraries/walkApi.js:
```diff
@@ -241,368 +503,417 @@ animation = (function() {
+                if (isDefined(sourceAnimation.filters[i])) {
+                    targetAnimation.filters[i] = sourceAnimation.filters[i];
+                    // wave shapers
+                    if (isDefined(sourceAnimation.filters[i].pitchFilter)) {
```
See isDefined comment above.

howard-stearns commented on examples/libraries/walkApi.js:
```diff
@@ -241,368 +503,417 @@ animation = (function() {
+                    if (isDefined(sourceAnimation.filters[i].pitchFilter)) {
+                        targetAnimation.filters[i].pitchFilter = sourceAnimation.filters[i].pitchFilter;
+                    }
+                    if (isDefined(sourceAnimation.filters[i].yawFilter)) {
```
See isDefined comment above.

howard-stearns commented on examples/libraries/walkApi.js:
```diff
@@ -241,368 +503,417 @@ animation = (function() {
+                    if (isDefined(sourceAnimation.filters[i].yawFilter)) {
+                        targetAnimation.filters[i].yawFilter = sourceAnimation.filters[i].yawFilter;
+                    }
+                    if (isDefined(sourceAnimation.filters[i].rollFilter)) {
```
See isDefined comment above.

howard-stearns commented on examples/libraries/walkApi.js:
```diff
@@ -241,368 +503,417 @@ animation = (function() {
+                        targetAnimation.filters[i].rollFilter = sourceAnimation.filters[i].rollFilter;
+                    }
+                    // LP filters
+                    if (isDefined(sourceAnimation.filters[i].swayLPFilter)) {
```
See isDefined comment above.

howard-stearns commented on examples/libraries/walkApi.js:
```diff
@@ -241,368 +503,417 @@ animation = (function() {
+                    if (isDefined(sourceAnimation.filters[i].swayLPFilter)) {
+                        targetAnimation.filters[i].swayLPFilter = sourceAnimation.filters[i].swayLPFilter;
+                    }
+                    if (isDefined(sourceAnimation.filters[i].bobLPFilter)) {
```
See isDefined comment above.

howard-stearns commented on examples/libraries/walkApi.js:
```diff
@@ -241,368 +503,417 @@ animation = (function() {
+                    if (isDefined(sourceAnimation.filters[i].bobLPFilter)) {
+                        targetAnimation.filters[i].bobLPFilter = sourceAnimation.filters[i].bobLPFilter;
+                    }
+                    if (isDefined(sourceAnimation.filters[i].thrustLPFilter)) {
```
See isDefined comment above.

howard-stearns commented on examples/libraries/walkApi.js:
```diff
@@ -241,368 +503,417 @@ animation = (function() {
+    // set parameters for the transition
+    this.parameters = new TransitionParameters();
+    this.liveReachPoses = [];
+    if (walkAssets && isDefined(lastAnimation) && isDefined(nextAnimation)) {
```
See isDefined comment above.

howard-stearns commented on examples/walk.js:
```diff
@@ -1,852 +1,500 @@
+var HIFI_PUBLIC_BUCKET = "https://hifi-public.s3.amazonaws.com/";
+
+// check for existence of data file object property
+function isDefined(value) {
```
See isDefined comment above.

howard-stearns commented on examples/walk.js:
```diff
@@ -1,852 +1,500 @@
-       (motion.direction === LEFT ||
-        motion.direction === RIGHT)) {
+    // if already transitioning from a blended walk need to maintain the previous walk's direction
+    if (isDefined(lastAnimation.lastDirection)) {
```
See isDefined comment above.

howard-stearns commented on examples/libraries/walkApi.js:
```diff
@@ -241,368 +503,417 @@ animation = (function() {
-        maleSideStepLeft: _maleSideStepLeft,
-        maleSideStepRight: _maleSideStepRight,
-        animationReference: _animationReference,
+    if (isDefined(joint)) {
```
See isDefined comment above.

howard-stearns commented on examples/libraries/walkApi.js:
```diff
@@ -241,368 +503,417 @@ animation = (function() {
-        maleSideStepRight: _maleSideStepRight,
-        animationReference: _animationReference,
+    if (isDefined(joint)) {
+        if (isDefined(joint.pitchFrequencyMultiplier)) {
```
See isDefined comment above.

howard-stearns commented on examples/libraries/walkApi.js:
```diff
@@ -241,368 +503,417 @@ animation = (function() {
+        if (isDefined(joint.pitchFrequencyMultiplier)) {
+            this.pitchFrequencyMultiplier = joint.pitchFrequencyMultiplier;
+        }
+        if (isDefined(joint.yawFrequencyMultiplier)) {
```
See isDefined comment above.

howard-stearns commented on examples/libraries/walkApi.js:
```diff
@@ -241,368 +503,417 @@ animation = (function() {
+        if (isDefined(joint.yawFrequencyMultiplier)) {
+            this.yawFrequencyMultiplier = joint.yawFrequencyMultiplier;
+        }
+        if (isDefined(joint.rollFrequencyMultiplier)) {
```
See isDefined comment above.

howard-stearns commented on examples/libraries/walkApi.js:
```diff
@@ -241,368 +503,417 @@ animation = (function() {
+        if (isDefined(joint.rollFrequencyMultiplier)) {
+            this.rollFrequencyMultiplier = joint.rollFrequencyMultiplier;
+        }
+        if (isDefined(joint.swayFrequencyMultiplier)) {
```
See isDefined comment above.

howard-stearns commented on examples/libraries/walkApi.js:
```diff
@@ -241,368 +503,417 @@ animation = (function() {
+        if (isDefined(joint.bobFrequencyMultiplier)) {
+            this.bobFrequencyMultiplier = joint.bobFrequencyMultiplier;
+        }
+        if (isDefined(joint.thrustFrequencyMultiplier)) {
```
See isDefined comment above.

howard-stearns commented on examples/walk.js:
```diff
@@ -1,852 +1,500 @@
+                setTransition(avatar.selectedHover, playTransitionReachPoses);
+            }
+            if (motion.state !== STATIC) {
+                motion.state = STATIC;
```
Any reason not to just always set motion.state = STATIC here?  No harm done if already STATIC, and less lines to grok.  Just a style question.

DaveDubUK replied:
There's no reason not to - I will update this.

howard-stearns commented on examples/walk.js:
```diff
@@ -1,852 +1,500 @@
+                    avatar.selectedWalkBlend.lastDirection = BACKWARDS;
+                    break;
+
+                case LEFT:
```
Do I understand correctly that this is for lateral (i.e., strafing) motion, rather than turning?

Years ago, I stole the Second Life turning .bvh for another system, and it worked pretty well (either alone, or blended with walking as appropriate).  Any thoughts of doing that here?

DaveDubUK replied:
Yep, this is just for lateral motion. Rotational motion (turning, turning on the spot) is planned for a future release.

Great to hear you've had success blending turning and walking animations, as I have had a plan to try this idea out for a while now. I also plan to experiment with blending a sidestep motion in too, as the Playstation controller (and others) can also generate a movement vectors that create a situation that needs this.

howard-stearns commented on examples/libraries/walkApi.js:
```diff
@@ -1,229 +1,491 @@
         var controllersPerTrigger = numberOfSpatialControls / numberOfTriggers;
-        hydrasConnected = (numberOfButtons == 12 && numberOfTriggers == 2 && controllersPerTrigger == 2);
-        return hydrasConnected;
+        if (numberOfButtons == 12 && numberOfTriggers == 2 && controllersPerTrigger == 2) {
```
Although our coding standard was really developed for C++, we've recently decided to try to use it for Javascript, too. http://docs.highfidelity.com/v1.0/docs/coding-standard#section-3-5-1-constants-and-magic-numbers says to not use magic numbers like this. (Although, admittedly, a lot of our javascript doesn't follow 3.5.1.2, and instead declares the consts at the top of the file.) Javascript const can be used instead of var.

I don't think its worth holding the merge for this, but if you happen to go into this part of the code...  Part of the idea with reviewing for coding standard adherence is that the code will certainly be looked at a lot going forward. (And this code especially!) So anything we can do to make it more uniformly readable now is going to pay off a lot with so many people looking at it later.

DaveDubUK replied:
Please do hold off the merge, I want to get one last version in first!

I'll put any remaining magic numbers into a named, const form.

howard-stearns commented on examples/libraries/walkApi.js:
```diff
@@ -1,229 +1,491 @@
+                this.selectedSideStepLeft = walkAssets.getAnimationDataFile("MaleSideStepLeft");
+                this.selectedSideStepRight = walkAssets.getAnimationDataFile("MaleSideStepRight");
+                this.selectedWalkBlend = walkAssets.getAnimationDataFile("WalkBlend");
+                this.selectedTurnLeft = walkAssets.getAnimationDataFile("MaleTurnLeft");
```
Is this line and the next ever made use of?

DaveDubUK replied:
No, not yet. I've already made a start on the turn on the spot animations, and forgot to remove the test code for the PR.

Will be removed.

howard-stearns commented on examples/libraries/walkApi.js:
```diff
@@ -1,229 +1,491 @@
-    this.direction = FORWARDS;
-    this.strideLength = this.selectedWalk.calibration.strideLengthForwards;
     this.frequencyTimeWheelPos = 0;
     this.frequencyTimeWheelRadius = 0.5;
```
magic number

howard-stearns commented on examples/libraries/walkApi.js:
```diff
@@ -1,229 +1,491 @@
     this.recentFrequencyTimeIncrements = [];
-    for(var i = 0; i < 8; i++) {
-
+    for (var i = 0; i < 8; i++) {
```
magic number

howard-stearns commented on examples/walk.js:
```diff
@@ -1,852 +1,500 @@
+Script.include([
+    "./libraries/walkFilters.js",
+    "./libraries/walkApi.js",
+    pathToAssets + "walkAssets.js"
```
I think walkAssets.js has the same sort of "library" issue as I mention for line 519 of walkAPI.js, in that walkAssets.js uses pathToAssets, which isn't defined in walkAssets.js.

DaveDubUK commented:
Thank you for all the feedback. I will make the changes described above and update.

howard-stearns commented on examples/libraries/walkApi.js:
```diff
@@ -1,229 +1,491 @@
+        };
+
+        if (this.nextStep === RIGHT) {
+            Audio.playSound(walkAssets.footsteps[0], options);
```
See my comments in https://worklist.net/20579

howard-stearns commented on examples/libraries/walkApi.js:
```diff
@@ -1,229 +1,491 @@
+            Audio.playSound(walkAssets.footsteps[0], options);
+            this.nextStep = LEFT;
+        } else if (this.nextStep === LEFT) {
+            Audio.playSound(walkAssets.footsteps[1], options);
```
Same audio issue. Create (in this case, two) injector(s), and reuse them, rather than creating new threads with repeated playSound.

DaveDubUK replied:
I tested with the setOptions call on the dev build, working well, so tried same method (minus the call to setOptions) on build 2701. Seems to be working great, motion smooth regardless of whether sound plays or not - thanks Howard!

There is one problem though, when Interface and the script are first started, the first call to Audio.playSound consistently produces a distorted sound - I made a recording: http://highfidelity.davedub.co.uk/bug-reports/audio-injector-startup-sound.wav

Note on linked sound above - it doesn't play in my browser (Chrome) for some reason, but if it's download it first, it will play fine.

howard-stearns commented:
I'll continue looking this in the next few days. However, we'll be pretty tied up tomorrow (Wednesday), and I'm away on Thursday.

DaveDubUK commented:
Thanks for all the help so far. I've addressed all your points. Please note that the icon / tab for the setting panel is not displaying in recent builds. I just tested on build 2649 and the icon / tab displays fine, so I've added a shortcut to access the settings panel (Shift O) as a workaround for current builds.

I've committed / pushed the changes to my walk-1.25 branch, which if I understand things right, should update the PR code.

hifi-gustavo commented:
Build console output for commit: 73e59fe07f8b5f72bfc5445c63a9b2d3809c8e30 available at: https://gist.github.com/9720a056bd33b39951a1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5179.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5179.dmg

hifi-gustavo commented:
Build console output for commit: 0e71583048ebea89c6dc253b77ec5bcb1dca59b0 available at: https://gist.github.com/9720a056bd33b39951a1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5179.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5179.dmg

hifi-gustavo commented:
Build console output for commit: ce31e586b0ab6ffbe2e180b48e41a8e3c3d463ae available at: https://gist.github.com/9720a056bd33b39951a1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5179.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5179.dmg

DaveDubUK commented:
Footstep sounds now working correctly and have been enabled by default for the first time :-)
Constants now in a separate file, accessible by both walk.js and walkApi.js, so walkApi.js can be used independently. 
walkAssets.js now has it's own path to assets defined, so can also be used independently of walk.js.

DaveDubUK commented:
Hi Howard,

I've had a good look at this audio issue. I can't see anything wrong with my code, but still the first sound after Interface has started up is always distorted (download wav clip here: http://highfidelity.davedub.co.uk/bug-reports/audio-injector-startup-sound.wav - note: you may need to download, as this won't play in my Chrome browser for some reason)
The footstep sounds seem to work fine after the initial glitch, but do seem to cut out every now and then after walking for a while.

I tried creating AudioInjectionOptions as per http://docs.highfidelity.com/v1.0/docs/audio-functions , but get the following error: ReferenceError: Can't find variable: AudioInjectionOptions.

Which means I'm out of options!

Am unsure how to proceed - I'd really like to get this last issue fixed, as I know the footstep sounds are high on Phillip's priorities. If you can shed any light on the situation, I'd be very grateful!

Cheers,

 - Dave

howard-stearns commented:
Your new audio code looks right. We'll have to take a separate look at the distortion.
I've merged.
Nice work!

DaveDubUK commented:
Hi Howard,

Thanks for your help Howard was great to work with you, and it's great to finally get my work synced!

However, the audio is enabled by default on this version, so everyone will get the distorted first sound - I think we should disable it. I have created a new PR with just this one change here: 
https://github.com/highfidelity/hifi/pull/5228

Please could you merge that for me?

Cheers!

Dave
